[SPARK-3218, SPARK-3219, SPARK-3261, SPARK-3424] [RESUBMIT] MLLIB K-Means Clusterer

derrickburns opened the pull request:
This commit introduces a general distance function trait, `PointOps`, for the Spark K-Means clusterer. There are no public API changes*.  
### Issue - Data Capture Test Fails - NEED HELP

The `org.apache.spark.mllib.clustering.KMeansClusterSuite` "task size should be small in both training and prediction" fails, suggesting that the RDD data is being captured in a closure.  This is quite puzzling. My efforts to solve this problem have failed. I _need help_ to solve this problem. 
### Distance Function Trait

The `PointOps` trait defines the distance function.  The `PointOps` trait is more than a simple distance function.  It also defines the types of Points and Centers for the clusterer.  Standard MLLIB `Vector`s are converted into Points and Centers.  In the case of the `FastEuclideanOps` implementation of `PointOps`, the Point and Center types includes vector norm members. In other distance functions such as the [Kullback-Leibler distance function](http://en.wikipedia.org/wiki/Bregman_divergence), the Point and Center types include different values that speed up the distance calculation in a similar way that caching vector norms speeds up the Euclidean distance function.  This addresses SPARK-3219.
### Refactoring

To understand this original code, I found it useful to refactor the original implementation into components.  You may find it helpful to understand this pull request by looking at the new components and comparing them to their original implementation.  Unfortunately, GitHub diff does not help very much with this.

This commit splits up the clusterer into a number of components which behave (largely) like their predecessors. `KMeansParallel` implements the K-Means || initialization algorithm.  `KMeansRandom` implements the K-Means Random initialization algorithm.  `MultiKMeans` implements the K-Means algorithm on multiple sets of cluster centers using a given distance function.  Traits for the initializer, `KMeansInitializer`, and the general K-Means clusterer, `MultiKMeansClusterer`, are provided to highlight the salient interfaces with the intent that alternative implementations of these interfaces may be provided in the future.
### Performance

This pull request is not focused on performance. Nevertheless,  the performance of the KMeans++ implementation was _dramatically_ improved by NOT recomputing distances to clusters centers that were present in previous steps.  This turns a quadratic implementation into a linear one. 

Second, the KMeans++ implementation uses the general K-Means clusterer in the final step.  This parallelizes a step that was sequential.

Together, these changes address SPARK-3424.
### Next Steps

This pull request does not introduce new user-visible changes.  The next step is to make different distance functions available through a user-visible API.   I will provide other distance functions after this pull request has been accepted. Then, we can decide on an appropriate user-level API to access those functions.
### Compatibility

While there are no user-level API changes, the behavior of the clusterer is _different_ on some tests.  Specifically, the handling of empty clusters has changed.  Empty clusters are not filled with random points in this implementation.  The former behavior is undesirable for a number a reasons, not the least of which is that there is no reasonable use for duplicate cluster centers. To accommodate the change in behavior, the test cases were changed accordingly. This addresses SPARK-3261.
### Testing

This code has been tested (albeit while packaged outside of Spark) and performance measured on data sets of millions of features each with hundreds of dimensions and on tens of thousands of clusters.

derrickburns committed:
This commit introduces a general distance function trait, PointOps, for the Spark K-Means clusterer. There are no public API changes*.

Distance Function Trait

The PointOps trait defines the distance function. The PointOps trait is more than a simple distance function. It also defines the types of Points and Centers for the clusterer. Standard MLLIB Vectors are converted into Points and Centers. In the case of the FastEuclideanOps implementation of PointOps, the Point and Center types includes vector norm members. In other distance functions such as the Kullback-Leibler distance function, the Point and Center types include different values that speed up the distance calculation in a similar way that caching vector norms speeds up the Euclidean distance function. This addresses SPARK-3219.

Refactoring

To understand this original code, I found it useful to refactor the original implementation into components. You may find it helpful to understand this pull request by looking at the new components and comparing them to their original implementation. Unfortunately, GitHub diff does not help very much with this.

This commit splits up the clusterer into a number of components which behave (largely) like their predecessors. KMeansParallel implements the K-Means || initialization algorithm. KMeansRandom implements the K-Means Random initialization algorithm. MultiKMeans implements the K-Means algorithm on multiple sets of cluster centers using a given distance function. Traits for the initializer, KMeansInitializer, and the general K-Means clusterer, MultiKMeansClusterer, are provided to highlight the salient interfaces with the intent that alternative implementations of these interfaces may be provided in the future.

Performance

This pull request is not focused on performance. Nevertheless, the performance of the KMeans++ implementation was dramatically improved by NOT recomputing distances to clusters centers that were present in previous steps. This turns a quadratic implementation into a linear one.

Second, the KMeans++ implementation uses the general K-Means clusterer in the final step. This parallelizes a step that was sequential.

Together, these changes address SPARK-3424.

Next Steps

This pull request does not introduce new user-visible changes. The next step is to make different distance functions available through a user-visible API. I will provide other distance functions after this pull request has been accepted. Then, we can decide on an appropriate user-level API to access those functions.

Compatibility

While there are no user-level API changes, the behavior of the clusterer is different on some tests. Specifically, the handling of empty clusters has changed. Empty clusters are not filled with random points in this implementation. The former behavior is undesirable for a number a reasons, not the least of which is that there is no reasonable use for duplicate cluster centers. To accommodate the change in behavior, the test cases were changed accordingly. This addresses SPARK-3261.

The private K-Means constructor which was used by some test Java code and one example was replaced with a Scala constructor that is not Java friendly. Since the constructor was not user visible, I simply changed the Java test code and the example to use the higher level interface.

Testing

This code has been tested (albeit while packaged outside of Spark) and performance measured on data sets of millions of features each with hundreds of dimensions and on tens of thousands of clusters.

derrickburns committed:
Fixed formatting to conform with Spark style guide.

derrickburns committed:
Restored KMeans class and setter methods.

derrickburns committed:
Spark coding formatting.

derrickburns committed:
Spark coding formatting.

derrickburns committed:
Added documentation.

derrickburns committed:
Use logInfo instead of log.info.

mengxr commented:
@derrickburns I tried to merge this PR with the current master. But there are many conflicts, majorly because we removed breeze from the implementation. To merge the changes, I think we should make separate PRs for the JIRAs mentioned here, from fixes to new features. Do you mind me splitting this PR? Thanks!

derrickburns commented:
I've said this before, so please forgive me for being repetitive.

The new implementation is a rewrite, not a patch, so it is not possible to
parcel out the changes into distinct PRs.

On Mon, Jan 5, 2015 at 1:09 PM, Xiangrui Meng <notifications@github.com>
wrote:

> @derrickburns <https://github.com/derrickburns> I tried to merge this PR
> with the current master. But there are many conflicts, majorly because we
> removed breeze from the implementation. To merge the changes, I think we
> should make separate PRs for the JIRAs mentioned here, from fixes to new
> features. Do you mind me splitting this PR? Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634#issuecomment-68781838>.
>

derrickburns commented:
The pull request that you integrated on December 3 is redundant to this
one. Therefore you need not worry about merging in those changes.  Simply
select this version.

Are you willing integrate this pull request as is?


On Mon, Jan 5, 2015 at 4:01 PM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> I've said this before, so please forgive me for being repetitive.
>
> The new implementation is a rewrite, not a patch, so it is not possible to
> parcel out the changes into distinct PRs.
>
> On Mon, Jan 5, 2015 at 1:09 PM, Xiangrui Meng <notifications@github.com>
> wrote:
>
>> @derrickburns <https://github.com/derrickburns> I tried to merge this PR
>> with the current master. But there are many conflicts, majorly because we
>> removed breeze from the implementation. To merge the changes, I think we
>> should make separate PRs for the JIRAs mentioned here, from fixes to new
>> features. Do you mind me splitting this PR? Thanks!
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/spark/pull/2634#issuecomment-68781838>.
>>
>
>

srowen commented:
(@derrickburns this PR can't be merged as-is. It contains merge conflicts with master now, as Github notes here. You would need to rebase it.)

mengxr commented:
@derrickburns I like the improvements implemented in this PR. But as @srowen mentioned, we have to resolve conflicts with the master branch before we can merge any PR. I compared the performance of this PR with master on minist-digits (60000x784, sparse, 10 clusters) locally and found the master runs 2-3x faster. I guess this is majorly caused by two changes.

1. We replaced breeze operations by our own implementation. The latter is about 2-3x faster.
1. Running k-means++ distributively has noticeable overhead with small k and feature dimension.

I think it is still feasible to include features through separate PRs:

1. remember previously computed best distances in k-means++ initialization
1. allow fixing the random seed (addressed in #3610)
1. variable number of clusters. We should discuss whether we want to have less than k clusters or split the biggest one if there are more than k points. 
1. parallelize k-means++. I think whether we should replace local k-means++ or make it configurable requires some discussion and performance comparison.
1. support Bregman divergences

Putting all of them together would certainly delay the review process and require resolving conflicts. I may have some time to prepare PRs for some of the features here, if you don't mind.

For Bregman divergences, I'm thinking we can alter the formulation to support sparse vectors:

~~~
d(x, y) = f(x)  - f(y) - <x - y, g(y)> = f(x) - (f(y) - <y, g(y)>) - <x, g(y)>
~~~

where `f(x)`, `g(y)`, and `f(y) - <y, g(y)>` could be pre-computed and cached, and `<x, g(y)>` can take advantage of sparse `x`. But I'm not sure whether this formulation is really useful on any Bregman divergence rather than the squared distance and the Mahalanobis distance. For KL-divergence and generalized I-divergence, the domain is R^d_+ and hence the points cannot be sparse.

Besides those comments, I'm going to make some minor comments inline.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala:
```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib
```
Should it be `mllib.clustering` as the file is under `clustering/`?

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala:
```diff
@@ -0,0 +1,77 @@
+import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
+
+class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
+  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
```
Should `weight` be only used in aggregation rather than distance computation?

derrickburns replied:
It is most efficient to maintain the cluster centers in homogeneous
coordinates.

On Thu, Jan 8, 2015 at 3:56 PM, Xiangrui Meng <notifications@github.com>
wrote:

> In
> mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala
> <https://github.com/apache/spark/pull/2634#discussion-diff-22693112>:
>
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib.clustering.metrics
> > +
> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
> > +
> > +import org.apache.spark.mllib.base._
> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
> > +
> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
>
> Should weight be only used in aggregation rather than distance
> computation?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634/files#r22693112>.
>

mengxr replied:
Could you elaborate more about the benefit of using homogeneous coordinates? It should be at least documented that what "weight" stands for here. If a cluster have 100 points, do you assign its center with weight 100? If so, how to compute the distance between this center and an input point?

derrickburns replied:
Simply put, computing the distance (and the cached values per point and centered used in the distance) is faster in homogeneous coordinates. 

Imagine computing dot(x,y), where x and y are vectors. If x and y are in homogeneous coordinates, then it is faster to perform the fit product in homogeneous coordinates then divide by the weights of the two vectors than it is to divide each homogenous coordinate by the weights first and the do the dot product.

Sent from my iPhone

> On Jan 19, 2015, at 5:02 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala:
> 
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib.clustering.metrics
> > +
> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
> > +
> > +import org.apache.spark.mllib.base._
> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
> > +
> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
> Could you elaborate more about the benefit of using homogeneous coordinates? It should be at least documented that what "weight" stands for here. If a cluster have 100 points, do you assign its center with weight 100? If so, how to compute the distance between this center and an input point?
> 
> —
> Reply to this email directly or view it on GitHub.
>

derrickburns replied:
For the squared Euclidean distance case, weights would naturally start at 1.0 for each input point, so that a cluster with 100 points would have weight 100.0.

However, for the KL divergence where the points are derived from frequencies, the initial weight could be the sum of the frequencies. This puts the points on the simplex without having to perform the division.

Sent from my iPhone

> On Jan 19, 2015, at 5:02 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala:
> 
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib.clustering.metrics
> > +
> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
> > +
> > +import org.apache.spark.mllib.base._
> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
> > +
> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
> Could you elaborate more about the benefit of using homogeneous coordinates? It should be at least documented that what "weight" stands for here. If a cluster have 100 points, do you assign its center with weight 100? If so, how to compute the distance between this center and an input point?
> 
> —
> Reply to this email directly or view it on GitHub.
>

derrickburns replied:
@mengxr

I spent some time thinking about clustering sparse data.  I realized that
there may be a need to embed the sparse data into lower dimensional
spaces.  Then, I realized that one of the Bregman divergences (i.e. the
generalized symmetrized Bregman divergence), also embeds data into a
different space before clustering.  This led me to the conclusion that
embedding the data into a convenient space just prior to clustering is a
very useful generic feature.  Adding it was trivial.  See the README on my
GitHub project for details.

I implemented two generic embeddings: 1) a sparse to dense embedding that
maps points isomorphically and 2) random indexing
<https://www.sics.se/~mange/papers/RI_intro.pdf> that maps high dimensional
data into lower dimension in a way that preserves similarity per the
Johnson-Lindenstrauss
lemma <http://en.wikipedia.org/wiki/Johnson%E2%80%93Lindenstrauss_lemma>. I
also recoded the embedding that I did for a generalized symmetrized KL
divergence to use the new embedding design.

I think that with these improvements one can check the "clusterer supports
sparse data" feature box in a practical sense.

I am about to being performance testing of the randoming indexing embedding
on large volumes of high dimensional sparse data.

Please take a look at the new PointOps trait definition.

On Mon, Jan 19, 2015 at 5:36 PM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> For the squared Euclidean distance case, weights would naturally start at
> 1.0 for each input point, so that a cluster with 100 points would have
> weight 100.0.
>
> However, for the KL divergence where the points are derived from
> frequencies, the initial weight could be the sum of the frequencies. This
> puts the points on the simplex without having to perform the division.
>
> Sent from my iPhone
>
> On Jan 19, 2015, at 5:02 PM, Xiangrui Meng <notifications@github.com>
> wrote:
>
> In
> mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala
> <https://github.com/apache/spark/pull/2634#discussion-diff-23197316>:
>
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib.clustering.metrics
> > +
> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
> > +
> > +import org.apache.spark.mllib.base._
> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
> > +
> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
>
> Could you elaborate more about the benefit of using homogeneous
> coordinates? It should be at least documented that what "weight" stands for
> here. If a cluster have 100 points, do you assign its center with weight
> 100? If so, how to compute the distance between this center and an input
> point?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634/files#r23197316>.
>
>

derrickburns replied:
@mengxr
I have rewritten the README file
<https://github.com/derrickburns/generalized-kmeans-clustering> for my
clustering GtHub project. I think that you may find it useful.

On Thu, Jan 22, 2015 at 7:43 PM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> @mengxr
>
> I spent some time thinking about clustering sparse data.  I realized that
> there may be a need to embed the sparse data into lower dimensional
> spaces.  Then, I realized that one of the Bregman divergences (i.e. the
> generalized symmetrized Bregman divergence), also embeds data into a
> different space before clustering.  This led me to the conclusion that
> embedding the data into a convenient space just prior to clustering is a
> very useful generic feature.  Adding it was trivial.  See the README on my
> GitHub project for details.
>
> I implemented two generic embeddings: 1) a sparse to dense embedding that
> maps points isomorphically and 2) random indexing
> <https://www.sics.se/~mange/papers/RI_intro.pdf> that maps high
> dimensional data into lower dimension in a way that preserves similarity
> per the Johnson-Lindenstrauss lemma
> <http://en.wikipedia.org/wiki/Johnson%E2%80%93Lindenstrauss_lemma>. I
> also recoded the embedding that I did for a generalized symmetrized KL
> divergence to use the new embedding design.
>
> I think that with these improvements one can check the "clusterer supports
> sparse data" feature box in a practical sense.
>
> I am about to being performance testing of the randoming indexing
> embedding on large volumes of high dimensional sparse data.
>
> Please take a look at the new PointOps trait definition.
>
> On Mon, Jan 19, 2015 at 5:36 PM, Derrick Burns <derrickrburns@gmail.com>
> wrote:
>
>> For the squared Euclidean distance case, weights would naturally start at
>> 1.0 for each input point, so that a cluster with 100 points would have
>> weight 100.0.
>>
>> However, for the KL divergence where the points are derived from
>> frequencies, the initial weight could be the sum of the frequencies. This
>> puts the points on the simplex without having to perform the division.
>>
>> Sent from my iPhone
>>
>> On Jan 19, 2015, at 5:02 PM, Xiangrui Meng <notifications@github.com>
>> wrote:
>>
>> In
>> mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala
>> <https://github.com/apache/spark/pull/2634#discussion-diff-23197316>:
>>
>> > + * distributed under the License is distributed on an "AS IS" BASIS,
>> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>> > + * See the License for the specific language governing permissions and
>> > + * limitations under the License.
>> > + */
>> > +
>> > +package org.apache.spark.mllib.clustering.metrics
>> > +
>> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
>> > +
>> > +import org.apache.spark.mllib.base._
>> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
>> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
>> > +
>> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
>> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
>>
>> Could you elaborate more about the benefit of using homogeneous
>> coordinates? It should be at least documented that what "weight" stands for
>> here. If a cluster have 100 points, do you assign its center with weight
>> 100? If so, how to compute the distance between this center and an input
>> point?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/spark/pull/2634/files#r23197316>.
>>
>>
>

derrickburns replied:
@mengxr
By separating the initialization of cluster centers from the clustering,
one may trivially implement several variants of the standard K-means
clustering, including the "Wavelet-Based Anytime Algorithm for K-Means
Clustering of Time Series",
<http://www.cs.gmu.edu/~jessica/publications/ikmeans_sdm_workshop03.pdf>
which has been demonstrated to give better quality clusterings in less time
than the standard algorithm.

To demonstrate this, I have implemented the aforementioned algorithm in
roughly 50 lines of new code. See
com.massivedatascience.clusterer.WaveletKMeans.

On Sat, Jan 24, 2015 at 1:45 AM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> @mengxr
> I have rewritten the README file
> <https://github.com/derrickburns/generalized-kmeans-clustering> for my
> clustering GtHub project. I think that you may find it useful.
>
> On Thu, Jan 22, 2015 at 7:43 PM, Derrick Burns <derrickrburns@gmail.com>
> wrote:
>
>> @mengxr
>>
>> I spent some time thinking about clustering sparse data.  I realized that
>> there may be a need to embed the sparse data into lower dimensional
>> spaces.  Then, I realized that one of the Bregman divergences (i.e. the
>> generalized symmetrized Bregman divergence), also embeds data into a
>> different space before clustering.  This led me to the conclusion that
>> embedding the data into a convenient space just prior to clustering is a
>> very useful generic feature.  Adding it was trivial.  See the README on my
>> GitHub project for details.
>>
>> I implemented two generic embeddings: 1) a sparse to dense embedding that
>> maps points isomorphically and 2) random indexing
>> <https://www.sics.se/~mange/papers/RI_intro.pdf> that maps high
>> dimensional data into lower dimension in a way that preserves similarity
>> per the Johnson-Lindenstrauss lemma
>> <http://en.wikipedia.org/wiki/Johnson%E2%80%93Lindenstrauss_lemma>. I
>> also recoded the embedding that I did for a generalized symmetrized KL
>> divergence to use the new embedding design.
>>
>> I think that with these improvements one can check the "clusterer
>> supports sparse data" feature box in a practical sense.
>>
>> I am about to being performance testing of the randoming indexing
>> embedding on large volumes of high dimensional sparse data.
>>
>> Please take a look at the new PointOps trait definition.
>>
>> On Mon, Jan 19, 2015 at 5:36 PM, Derrick Burns <derrickrburns@gmail.com>
>> wrote:
>>
>>> For the squared Euclidean distance case, weights would naturally start
>>> at 1.0 for each input point, so that a cluster with 100 points would have
>>> weight 100.0.
>>>
>>> However, for the KL divergence where the points are derived from
>>> frequencies, the initial weight could be the sum of the frequencies. This
>>> puts the points on the simplex without having to perform the division.
>>>
>>> Sent from my iPhone
>>>
>>> On Jan 19, 2015, at 5:02 PM, Xiangrui Meng <notifications@github.com>
>>> wrote:
>>>
>>> In
>>> mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala
>>> <https://github.com/apache/spark/pull/2634#discussion-diff-23197316>:
>>>
>>> > + * distributed under the License is distributed on an "AS IS" BASIS,
>>> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>>> > + * See the License for the specific language governing permissions and
>>> > + * limitations under the License.
>>> > + */
>>> > +
>>> > +package org.apache.spark.mllib.clustering.metrics
>>> > +
>>> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
>>> > +
>>> > +import org.apache.spark.mllib.base._
>>> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
>>> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
>>> > +
>>> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
>>> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
>>>
>>> Could you elaborate more about the benefit of using homogeneous
>>> coordinates? It should be at least documented that what "weight" stands for
>>> here. If a cluster have 100 points, do you assign its center with weight
>>> 100? If so, how to compute the distance between this center and an input
>>> point?
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/apache/spark/pull/2634/files#r23197316>.
>>>
>>>
>>
>

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala:
```diff
@@ -0,0 +1,77 @@
+    val d = if (p.weight == Zero || c.weight == Zero) {
+      p.norm + c.norm
+    } else {
+      val x = p.raw.dot(c.raw) / (p.weight * c.weight)
```
same question about using `weight` in `distance`

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala:
```diff
@@ -0,0 +1,145 @@
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.rdd.RDD
+
+package object base {
```
Though private, having some doc would make the code easy to understand. For example,

1. What does `FP` mean?
1. `inh`?
1. ...

derrickburns replied:
I'm willing to take one last crack at contributing the Bregman modifications to Spark. Here is an layering of the pull requests:

1) modifications to BLAS to support arbitrary functions and axpy on SparseVectors
2) definition of PointOps trait and various implementations for different divergences 
3) rewrite of clusterer to use PointOps trait but only using the squared Euclidean distance implementation 
4) generalization of clustering interface to select different divergences.
5) store seeds
6) maintain exactly K clusters by splitting clusters heuristicly.

I can submit PRs for 1-4. 

Thoughts?

Sent from my iPhone

> On Jan 8, 2015, at 4:00 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala:
> 
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib
> > +
> > +import breeze.linalg.{ Vector => BV }
> > +
> > +import org.apache.spark.mllib.linalg.Vector
> > +import org.apache.spark.rdd.RDD
> > +
> > +package object base {
> Though private, having some doc would make the code easy to understand. For example,
> 
> What does FP mean?
> inh?
> ...
> —
> Reply to this email directly or view it on GitHub.
>

derrickburns replied:
@mengxr

I tried to build the Spark master branch, but was unsuccessful.

So, instead of integrating my Bregman clusterer into the latest master
branch of Spark, I integrated them into my public GitHub project
<https://github.com/derrickburns/generalized-kmeans-clustering>.

Per your suggestion, this   version supports sparse vectors.

Included is support for several Bregman divergences including: squared
Euclidean distance, Kullback Leibler divergence, logistic loss, and
generalized I-divergence.

I created a general BregmanDivergence trait and a separate BregmanPointOps
trait.  When mixed in together, one gets a specific BregmanPointOps object
that computes Bregman divergences very efficiently, caching one value per
point and two values per center (per your observation).  This version is
easier to understand that the previous one.

Included also are three different implementations of K-Means clustering: 1)
one that caches distances (CachingKMeans) 2) one that tracks which clusters
moved and which points are assigned to which clusters and the distance to
the closest cluster (TrackingKMeans); and, 3) one that recomputes all
clusters and distances (SingleKMeans).

I have found the the TrackingKMeans clusterer has the best performance on
last data sets.

I have also found that maintaining cluster centers in homogeneous
coordinates is more efficient than using inhomogeneous coordinates for
moderate to high dimensional data. There is no penalty for low dimensional
data.

If/when you include such changes in Spark, I would appreciate being
included in the list of contributors.

Regards,

Derrick Burns



On Sun, Jan 11, 2015 at 10:52 AM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> I'm willing to take one last crack at contributing the Bregman
> modifications to Spark. Here is an layering of the pull requests:
>
> 1) modifications to BLAS to support arbitrary functions and axpy on
> SparseVectors
> 2) definition of PointOps trait and various implementations for different
> divergences
> 3) rewrite of clusterer to use PointOps trait but only using the squared
> Euclidean distance implementation
> 4) generalization of clustering interface to select different divergences.
> 5) store seeds
> 6) maintain exactly K clusters by splitting clusters heuristicly.
>
> I can submit PRs for 1-4.
>
> Thoughts?
>
> Sent from my iPhone
>
> On Jan 8, 2015, at 4:00 PM, Xiangrui Meng <notifications@github.com>
> wrote:
>
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala
> <https://github.com/apache/spark/pull/2634#discussion-diff-22693266>:
>
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib
> > +
> > +import breeze.linalg.{ Vector => BV }
> > +
> > +import org.apache.spark.mllib.linalg.Vector
> > +import org.apache.spark.rdd.RDD
> > +
> > +package object base {
>
> Though private, having some doc would make the code easy to understand.
> For example,
>
>    1. What does FP mean?
>    2. inh?
>    3. ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634/files#r22693266>.
>
>

mengxr replied:
> 1) modifications to BLAS to support arbitrary functions and axpy on SparseVectors

We already support axpy on SparseVectors in `mllib.linalg.BLAS`. What do you mean by `arbitrary functions`? If the target is to improve k-means, we only need to implement BLAS ops required by k-means.

> 2) definition of PointOps trait and various implementations for different divergences
> 3) rewrite of clusterer to use PointOps trait but only using the squared Euclidean distance implementation
> 4) generalization of clustering interface to select different divergences.

Sounds good.

> 5) store seeds

This is already covered in https://github.com/apache/spark/pull/3610, which is almost ready to merge.

> 6) maintain exactly K clusters by splitting clusters heuristicly.

We had an example in `StreamingKMeans`. But if we want to support different distances, we need to consider how to keep the new centers inside the domain.

Thanks for improving your implementation! Before we merge features one by one, it would be nice to maintain your implementation as a 3rd-party Spark package (spark-packages.org). That helps people find it and send feedbacks.

derrickburns commented:
Thanks for the information on the speedup that you obtained by eliminating
Breeze.  I was unaware that the performance is so poor.  To what do you
attribute the poor performance of Breeze?

One case certainly split out 1-4 AFTER doing 5. My point was that doing 5
(supporting Bregman divergences) requires touching practically every method
in the clustering package.

3. Supporting a variable number of clusters requires a significant rewrite
at well, since the assumption that there are K clusters is made repeatedly
in the code.  If seems like adding a way to split clusters introduces a
whole other dimension of flexibility -- which I am not against -- that
would complicate the interface.  For example, what heuristic would you use
to identify a cluster to split and how would you split it.  It seems to me
that once you go down that path, you are headed toward a divisive
clustering algorithm.

4. As for the overhead of running k-means++ distributively, will someone
use Spark for small k and feature dimension?

5. Bregman divergences

As it turns out, I now have a need to perform clustering on sparse data, so
I am making a pass over my private implementation to 1) eliminate the use
of Breeze and 2) use the Spark Vector classes.   When I am done, I can
include these modification in an updated pull request.

I do not understand you conclusion that for KL-divergence and generalized
I-divergence the points cannot be sparse.  The domain need not be R^d+.
Rather, the domain can be R^infinity+.

For example, in the problem that I am currently working on, I am creating
sparse vectors where the vectors represent the frequency of occurrence of
the contexts for given n-grams.  The space of the contexts is extremely
large, however, I am only interested in the M most frequent contexts per
n-gram.  I use a sketch to identify the M most frequent contexts per
n-gram.  This gives me an M length SparseVector of non-negative frequency
values.  Thus, for each n-gram, I will have a SparseVector that represents
the distribution of the more frequently occurring contexts.  I do not see a
problem using this SparseVector with the KL-divergence metric. Am I missing
something?

On Thu, Jan 8, 2015 at 3:55 PM, Xiangrui Meng <notifications@github.com>
wrote:

> @derrickburns <https://github.com/derrickburns> I like the improvements
> implemented in this PR. But as @srowen <https://github.com/srowen>
> mentioned, we have to resolve conflicts with the master branch before we
> can merge any PR. I compared the performance of this PR with master on
> minist-digits (60000x784, sparse, 10 clusters) locally and found the master
> runs 2-3x faster. I guess this is majorly caused by two changes.
>
>    1. We replaced breeze operations by our own implementation. The latter
>    is about 2-3x faster.
>    2. Running k-means++ distributively has noticeable overhead with small
>    k and feature dimension.
>
> I think it is still feasible to include features through separate PRs:
>
>    1. remember previously computed best distances in k-means++
>    initialization
>    2. allow fixing the random seed (addressed in #3610
>    <https://github.com/apache/spark/pull/3610>)
>    3. variable number of clusters. We should discuss whether we want to
>    have less than k clusters or split the biggest one if there are more than k
>    points.
>    4. parallelize k-means++. I think whether we should replace local
>    k-means++ or make it configurable requires some discussion and performance
>    comparison.
>    5. support Bregman divergences
>
> Putting all of them together would certainly delay the review process and
> require resolving conflicts. I may have some time to prepare PRs for some
> of the features here, if you don't mind.
>
> For Bregman divergences, I'm thinking we can alter the formulation to
> support sparse vectors:
>
> d(x, y) = f(x)  - f(y) - <x - y, g(y)> = f(x) - (f(y) - <y, g(y)>) - <x, g(y)>
>
> where f(x), g(y), and f(y) - <y, g(y)> could be pre-computed and cached,
> and <x, g(y)> can take advantage of sparse x. But I'm not sure whether
> this formulation is really useful on any Bregman divergence rather than the
> squared distance and the Mahalanobis distance. For KL-divergence and
> generalized I-divergence, the domain is R^d_+ and hence the points cannot
> be sparse.
>
> Besides those comments, I'm going to make some minor comments inline.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634#issuecomment-69272196>.
>

derrickburns commented:
I see the problem with sparse vectors and the KL divergence. 

I implemented a smoothing operation to approximate KL divergence.

Sent from my iPhone

> On Jan 8, 2015, at 3:55 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> @derrickburns I like the improvements implemented in this PR. But as @srowen mentioned, we have to resolve conflicts with the master branch before we can merge any PR. I compared the performance of this PR with master on minist-digits (60000x784, sparse, 10 clusters) locally and found the master runs 2-3x faster. I guess this is majorly caused by two changes.
> 
> We replaced breeze operations by our own implementation. The latter is about 2-3x faster.
> Running k-means++ distributively has noticeable overhead with small k and feature dimension.
> I think it is still feasible to include features through separate PRs:
> 
> remember previously computed best distances in k-means++ initialization
> allow fixing the random seed (addressed in #3610)
> variable number of clusters. We should discuss whether we want to have less than k clusters or split the biggest one if there are more than k points.
> parallelize k-means++. I think whether we should replace local k-means++ or make it configurable requires some discussion and performance comparison.
> support Bregman divergences
> Putting all of them together would certainly delay the review process and require resolving conflicts. I may have some time to prepare PRs for some of the features here, if you don't mind.
> 
> For Bregman divergences, I'm thinking we can alter the formulation to support sparse vectors:
> 
> d(x, y) = f(x)  - f(y) - <x - y, g(y)> = f(x) - (f(y) - <y, g(y)>) - <x, g(y)>
> where f(x), g(y), and f(y) - <y, g(y)> could be pre-computed and cached, and <x, g(y)> can take advantage of sparse x. But I'm not sure whether this formulation is really useful on any Bregman divergence rather than the squared distance and the Mahalanobis distance. For KL-divergence and generalized I-divergence, the domain is R^d_+ and hence the points cannot be sparse.
> 
> Besides those comments, I'm going to make some minor comments inline.
> 
> —
> Reply to this email directly or view it on GitHub.
>

SparkQA commented:
[QA tests have started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25711/consoleFull) for   PR 2634 at commit [`35da8e9`](https://github.com/apache/spark/commit/35da8e9e188e669460000d5799d061ecc3ca150f).
 * This patch **does not** merge cleanly!

SparkQA commented:
[QA tests have finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25711/consoleFull) for   PR 2634 at commit [`35da8e9`](https://github.com/apache/spark/commit/35da8e9e188e669460000d5799d061ecc3ca150f).
 * This patch **fails** unit tests.
 * This patch **does not** merge cleanly!

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25711/
Test FAILed.

derrickburns commented:
@mengxr

I have implemented several variants of Kullback-Leibler divergence in
my separate
GitHub repository
<https://github.com/derrickburns/generalized-kmeans-clustering>.  These
variants are more efficient that the standard KL-divergence which is
defined on R+ ^ n because they take advantage of extra knowledge of the
domain. I have used these variants with much success (i.e. much faster
running time) in my large scale clustering runs.

On Sat, Jan 17, 2015 at 7:02 PM, UCB AMPLab <notifications@github.com>
wrote:

> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25711/
> Test FAILed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634#issuecomment-70394598>.
>

derrickburns commented:
@mengxr

One more thing regarding sparse vectors.  Sparse vectors can become dense
under cluster creation, which, in turn, can cause the running time of the
K-means clustering to skyrocket.

To address this problem, one can project clusters onto a sparse vector
before performing distance calculation.  My current version of the
clusterer does this when the appropriate distance object is selected.

On Sun, Jan 18, 2015 at 7:59 PM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> @mengxr
>
> I have implemented several variants of Kullback-Leibler divergence in my separate
> GitHub repository
> <https://github.com/derrickburns/generalized-kmeans-clustering>.  These
> variants are more efficient that the standard KL-divergence which is
> defined on R+ ^ n because they take advantage of extra knowledge of the
> domain. I have used these variants with much success (i.e. much faster
> running time) in my large scale clustering runs.
>
> On Sat, Jan 17, 2015 at 7:02 PM, UCB AMPLab <notifications@github.com>
> wrote:
>
>> Test FAILed.
>> Refer to this link for build results (access rights to CI server needed):
>> https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25711/
>> Test FAILed.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/spark/pull/2634#issuecomment-70394598>.
>>
>
>

mengxr commented:
@derrickburns We use dense vectors to store cluster centers, because the centers are very likely to become dense during aggregation. If there are zeros, they can be efficiently compressed before sending back to the driver. For performance, we never add two sparse vectors together. I'm not sure whether this answers your question.

derrickburns commented:
In my application (n-gram contexts), the sparse vectors can be of extremely high dimension. To make the problem manageable, I select the k most important dimensions per point. For a cluster of m points, I can have a sparse vector of m*k non-zero values. Since some clusters can become quite large ( O(n) is size), I can get sparse vectors of O(nk) non-zero values.  Still, the vector is sparse, since the dimension of the vector is potentially 2^31-1.

So, I cannot treat the vector as dense. 

To deal with the growth problem, I have implemented a centroid that only retains the k most important features.

Sent from my iPhone

> On Jan 19, 2015, at 5:07 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> @derrickburns We use dense vectors to store cluster centers, because the centers are very likely to become dense during aggregation. If there are zeros, they can be efficiently compressed before sending back to the driver. For performance, we never add two sparse vectors together. I'm not sure whether this answers your question.
> 
> —
> Reply to this email directly or view it on GitHub.
>

srowen commented:
Do you mind closing this PR?

derrickburns commented:
Sure, but I'm traveling now. Would you mind closing it for me?

Sent from my iPhone

> On May 18, 2015, at 8:19 PM, Sean Owen <notifications@github.com> wrote:
> 
> Do you mind closing this PR?
> 
> —
> Reply to this email directly or view it on GitHub.
>

srowen commented:
We can't directly but there's an automated process that will eventually. Don't worry about it and/or get to it later.

derrickburns commented:
Ok thx !

Sent from my iPhone

> On May 18, 2015, at 8:47 PM, Sean Owen <notifications@github.com> wrote:
> 
> We can't directly but there's an automated process that will eventually. Don't worry about it and/or get to it later.
> 
> —
> Reply to this email directly or view it on GitHub.
>
